### PR TITLE
Messages refactor

### PR DESCRIPTION
**Pull request by VictorMikheev**

### CODE REVIEW

File: examples/src/main/scala/examples/commons/PublicKey25519NoncedBox.scala [hunk 1]
```diff
@@ -39,19 +33,21 @@ object PublicKey25519NoncedBox extends ScorexEncoding {
...
+object PublicKey25519NoncedBoxSerializer extends ScorexSerializer[PublicKey25519NoncedBox] {
 
-  override def toBytes(obj: PublicKey25519NoncedBox): Array[Byte] =
-    obj.proposition.pubKeyBytes ++
-      Longs.toByteArray(obj.nonce) ++
-      Longs.toByteArray(obj.value)
 
+  override def serialize(obj: PublicKey25519NoncedBox, w: Writer): Unit = {
+    PublicKey25519PropositionSerializer.serialize(obj.proposition, w)
+    w.putULong(obj.nonce)
```
Comment (kushti) [inline examples/src/main/scala/examples/commons/PublicKey25519NoncedBox.scala]: It seems the nonce could be negative (it is generated by applying Longs.fromByteArray to pseudorandom bytes).

File: examples/src/main/scala/examples/commons/SimpleBoxTransaction.scala [hunk 2]
```diff
@@ -142,42 +154,46 @@ object SimpleBoxTransaction extends ScorexEncoding {
...
+    w.putULong(m.timestamp)
+    w.putUInt(m.signatures.length)
+    w.putUInt(m.from.length)
+    w.putUInt(m.to.length)
+    m.signatures.foreach( s =>
+      Signature25519Serializer.serialize(s, w)
     )
+    m.from.foreach { f =>
+      PublicKey25519PropositionSerializer.serialize(f._1, w)
+      w.putULong(f._2)
```
Comment (kushti) [inline examples/src/main/scala/examples/commons/SimpleBoxTransaction.scala]: Again, Nonce could be negative.

File: examples/src/main/scala/examples/hybrid/blocks/PowBlock.scala [hunk 3]
```diff
@@ -29,40 +27,41 @@ class PowBlockHeader(
...
-    val brothersCount = Ints.fromByteArray(bytes.slice(80, 84))
-    val brothersHash = bytes.slice(84, 116)
-    val prop = PublicKey25519Proposition(PublicKey @@ bytes.slice(116, 148))
+object PowBlockHeaderSerializer extends ScorexSerializer[PowBlockHeader] {
+
+  override def serialize(h: PowBlockHeader, w: Writer): Unit = {
+     w.putBytes(idToBytes(h.parentId))
+     w.putBytes(idToBytes(h.prevPosId))
+     w.putULong(h.timestamp)
+     w.putULong(h.nonce)
```
Comment (kushti) [inline examples/src/main/scala/examples/hybrid/blocks/PowBlock.scala]: Probably, nonce could be negative here as well.

**Comment by kushti** [comment]
@VictorMikheev I've found few place where putULong is used for potentially negative number. Please review the whole PR for issues like that!

**Comment by catena2w** [comment]
Mostly included in https://github.com/ScorexFoundation/Scorex/pull/329

**Pull request closed without merging**